src/components/tl/ucp/reduce_scatter/reduce_scatter_knomial.c

```diff
@@ -180,25 +178,42 @@ void ucc_tl_ucp_reduce_scatter_knomial_progress(ucc_coll_task_t *coll_task)
                            "failed to perform dt reduction",
                            task->reduce_scatter_kn.etask);
         }
-        ucc_knomial_pattern_next_iteration(p);
+        ucc_kn_rs_pattern_next_iter(p);
     }
 
     if (!task->reduce_scatter_kn.scratch_mc_header) {
+        ucc_knomial_pattern_prev_iteration(p);
+        get_sbuf_rbuf(p, args, task->reduce_scatter_kn.scratch, block_count,
+                      &sbuf, &rbuf);
+
+        step_radix = ucc_kn_compute_step_radix(p);
+        local_seg_index = ucc_kn_compute_seg_index(
+            rank, p->radix_pow, p);
+        local_seg_count = ucc_sra_kn_compute_seg_size(
+            block_count, step_radix, local_seg_index);
+        local_seg_offset = ucc_sra_kn_compute_seg_offset(
+            block_count, step_radix, local_seg_index);
+        local_data  = PTR_OFFSET(sbuf, local_seg_offset * dt_size);
+        is_avg      = args->op == UCC_OP_AVG &&
+                            (avg_pre_op ? ucc_knomial_pattern_loop_first_iteration(p)
+                                        : ucc_knomial_pattern_loop_last_iteration(p));
+
         ucc_sra_kn_get_offset_and_seglen(count, dt_size, rank, size, radix,
                                          &offset, &local_seg_count);
-        eargs.task_type = UCC_EE_EXECUTOR_TASK_COPY;
-        eargs.copy.dst  = PTR_OFFSET(args->dst.info.buffer, offset);
-        eargs.copy.src  = task->reduce_scatter_kn.scratch;
-        eargs.copy.len  = local_seg_count * dt_size;
-        status = ucc_ee_executor_task_post(task->reduce_scatter_kn.executor, &eargs,
-                                           &task->reduce_scatter_kn.etask);
+        status = ucc_dt_reduce(local_data, rbuf,
```

Comment:
I seem to misunderstand the change. So, in case of !scratch (full pow radix case + no INPLACE) on the last reduction you do it in 2 steps: first you add all the buffers received during last step, place result in the beginning of the "rbuf" (which is dst.buffer). Then you do 1 more step in the end: reduce local data with that intermediate result and place the answer right back rbuf. Right?

How is that different from doing the entire reduce_strided with step_radix - 1?

Reply:
yes, it's correct, the difference is that in this PR we do step_radix -1 reductions while in prev version it was step_radix -1 reduction plus memory copy
i see big improvement for cases where this memory copy takes a lot of time for example 
![image](https://user-images.githubusercontent.com/5495223/228525603-b360da9d-6617-44e0-bbe0-377cfc1b527f.png)